pep-0626.rst

```diff
@@ -182,6 +182,53 @@ Access to the ``f_lineno`` attribute of frame objects through C API functions is
 
 Accessing ``f_lineno`` directly through the underlying data structure is forbidden.
 
+Out of process debuggers and profilers
+''''''''''''''''''''''''''''''''''''''
+
+Out of process tools, such as py-spy [1]_, cannot use the C-API, and must parse the line number table themselves.
+To reduce the work required to implement these tools, the following C struct and functions are provided.
+
+::
+
+    typedef struct addressrange {
+        int ar_start;
+        int ar_end;
+        int ar_line;
+        int opaque1;
+        void *opaque2;
+    } PyCodeAddressRange;
+
+    void PyLineTable_InitAddressRange(char *linetable, int firstlineno, PyCodeAddressRange *range);
+    int PyLineTable_NextAddressRange(PyCodeAddressRange *range);
+    int PyLineTable_PreviousAddressRange(PyCodeAddressRange *range);
+
+``PyLineTable_InitAddressRange`` initializes the ``PyCodeAddressRange`` struct from the line number table and first line number.
+
+``PyLineTable_NextAddressRange`` advances the range to the next entry, returning non-zero if valid.
+
+``PyLineTable_PreviousAddressRange`` retreats the range to the previous entry, returning non-zero if valid.
+
+For example, the following code prints out all the address ranges:
+
+::
+
+    void print_address_ranges(char *linetable, int firstlineno)
```

Comment:
This won't work as `py-spy` cannot execute code in the process being traced

Reply:
That doesn't make sense to me. 
`py-spy` must execute code; it's software.

Or are you saying that `py-spy` cannot use `PyLineTable_InitAddressRange`, etc. for licensing reasons, or similar?

Reply:
Presumably py-spy is something like gdb attaching to a different process. It can easily read any data in the process, but it is not so easy to call a function. (I think gdb can, but I presume it requires borrowing a thread, and given that py-spy is a profiling tool I think it would rather not do so.)

Reply:
Exactly: these tools attach to a process using `ptrace` and only read memory from it using the `ptrace` or `process_vm_readv` system calls. They will not execute code in the remote process due to security concerns (and also because is orders of magnitude more difficult than just reading memory). The setup is basically you have process A running Python and process B running the profiler. Process B can be implemented in any other language and will not be linked against `libpython.so`. Then process B attaches to process A and reads its memory (the location of the memory and the layout are obtaining using DWARF information or hardcoded offsets, there are many techniques). 

These tools can also analyze core files to get the python stack trace by reading the core file memory and in these core files, you cannot execute functions either (because is a core file not a living process).

Reply:
I believe you that these tools can't execute code in a different process.

But why can't they execute code in their own process, like all other programs? That's the bit I don't get.

Reply:
> But why can't they execute code in their own process, like all other programs? That's the bit I don't get.

How they will execute `PyLineTable_NextAddressRange`? These programs do not link against any `libpython` and can be implemented in any language. You say:

> To reduce the work required to implement these tools, the following C struct and functions are provided.

but when you mention "are provided"...where is this provided? If this is provided as part of the C-API then it won't work because these programs do not link against libpython (and even if they do, they need to analyze multiple versions of python so they would need to link against multiple different libpythons simultaneously which is not possible). Or are you  meaning that these programs should implement analogous of those functions on their own?

Reply:
Yeah, I can’t easily call python c-api functions from within py-sy =(.  Since py-spy copies memory from the profiled python process, that means any pointers we read won’t point to valid memory in the py-spy process - so passing these copied structs to libpython would just segfault right away.  Py-spy also supports multiple versions of python from a single py-spy binary which would make linking harder.

I’d rather not have py-spy run code in the profiled process. Its [technically possible](https://eklitzke.org/ptrace-continued), but modifying the target process like this is dangerous, and the steps are different on each OS/ CPU architecture. 

Fwiw, I’m perfectly happy to implement new code with py-spy to handle the new line format - I don’t want py-spy compatibility to slow down core python development =). However that will involve parsing the new co_linetable attribute directly - and I saw here ```the format is opaque, unspecified and may be changed without notice``` which sounds like this is discouraged.

Reply:
Please stop with all the stuff about running code in another process. It's irrelevant.
I know that an out of process debugger can't (easily) use the API of the running Python process, it says so in this PR.
https://github.com/python/peps/pull/1555/files#diff-e5a7e9acca6a66fb142d5a627bd16cabR188

You can run `PyLineTable_NextAddressRange` is any process you like. Run it in the debugger process; run it in the Python process; run it in a webapp on a Raspberry Pi, if you want. It doesn't matter.
`PyLineTable_NextAddressRange` is portable C code, you can run it anywhere you like.

Regarding the format. Yes, it might be changed without notice, which is why the `PyLineTable` functions are provided. 
I will clarify that the format won't change during a release, unless absolutely necessary for a bug fix.

So, just use the `PyLineTable` functions and you don't need to worry about the internal format. Copy the data across, as is already done for the `co_lnotab` data, and use the provided functions.

`py-spy`, and AFAICT other profilers and debuggers, need additional code for each release, so linking in the latest version of the `PyLineTable` functions shouldn't be a problem. Just make your own dll/so for each version and load that once you know which version of Python you're attached to.

Reply:
> So, just use the PyLineTable functions and you don't need to worry about the internal format. Copy the data across, as is already done for the co_lnotab data, and use the provided functions.

Thanks for the explanation! If I'm understanding you, I think what you're proposing could work for py-spy - but I have some questions:

1) Can I treat co_linetable like a PyVarObject to figure out the number of bytes to copy? The text in this PEP made it sound like co_linetable was an opaque pointer and wasn't clear that I could  get the size of the struct. I don't need the exact format, but do need to know how many bytes to copy over.

2) Will the PyLineTable functions access any pointer values inside of co_linetable? If they are only accessing data stored inside co_linetable, and don't follow any pointers from that struct we should be good. Otherwise we need to add a callback to do the memory access to the PyLineTable API's (any pointers will point to the memory space of the profiled program - so accessing in py-spy will segfault).

3) Will the PyLineTable functions have any other dependencies on other code in the python runtime? If I can just include the C source for only these functions and then bundle with py-spy that would work pretty well. I can't use dlopen to load a dll/so built per python version like you suggested (py-spy statically links to musl libc on linux), but can statically link in these functions by mangling the names or wrapping in a c++ namespace easily enough.

Btw, thanks for thinking about py-spy when designing this - I really appreciate it .

Reply:
> 3\. Will the PyLineTable functions have any other dependencies on other code in the python runtime? If I can just include the C source for only these functions and then bundle with py-spy that would work pretty well.

I think this is the best approach, like many other tools other than `py-spy` may have further constrains. For instance, if one of these tools is already linked against a given Python runtime (I maintain such a tool myself that for instance is implemented itself in Python as a C extension) it won't be able to link against any other runtimes so including the source and manging it will be verbatim. Given the new clarification:

> Note that these functions are not part of the C-API, so will be need to be linked into any code that needs to use them.

I assume this means that these functions won't need to use the C-API so IIUC vendoring the source code of these functions and operating over a copied blob of data will be enough.

Reply:
> 1. Can I treat co_linetable like a PyVarObject to figure out the number of bytes to copy? The text in this PEP made it sound like co_linetable was an opaque pointer and wasn't clear that I could  get the size of the struct. I don't need the exact format, but do need to know how many bytes to copy over.

I assume `co_linetable` must be either a Python object (in which case you can get the size as if it were a `PyVarObject `), a structure (in which case you can use DWARF info, a hardcoded offset or `offsetof` over a vendored copy of the struct or an opaque pointer + a size parameter somewhere.

Reply:
`co_linetable` is a Python object, for now.
The layout of code objects is not part of the C-API, so might well change in future versions.

Reply:
> 1. Can I treat co_linetable like a PyVarObject to figure out the number of bytes to copy?

Yes, for now. It may change in future versions.

> 2. Will the PyLineTable functions access any pointer values inside of co_linetable?

No
> 3. Will the PyLineTable functions have any other dependencies on other code in the python runtime?

No.

Reply:
Note that `PyLineTable_InitAddressRange` expects a `char *. It knows nothing about object layout.

Reply:
Just to clarify, the `PyLineTable functions` will only access memory inside the array of bytes passed to `PyLineTable_InitAddressRange`, but nothing outside of that array. It knows nothing of the original code object.

Reply:
awesome - that sounds good thanks!